docs/python-api.rst

```diff
@@ -374,6 +374,61 @@ see that the first metadata value is 9 bytes long and the second is 24.
 The :func:`tskit.pack_bytes` and :func:`tskit.unpack_bytes` functions are
 also useful for encoding data in these columns.
 
+.. _sec_tables_api_binary_columns_metadata:
+
+Metadata schemas
+################
+
+When a ``metadata_schema`` is present on the table the ``metadata`` binary column can be
+validated, encoded and decoded by tskit. Repeating the example above, but this time
+with a schema::
+
+    >>> table = tskit.NodeTable()
+    >>> table.metadata_schema = tskit.metadata.JSONMetadataSchema(
+        {
+            "type": "object",
+            "properties": {
+                "accession_number": {"type": "number"},
+                "collection_date": {
+                    "name": "Collection date",
+                    "description": "Date of sample collection in ISO format",
+                    "type": "string",
+                    "pattern": "^([1-9][0-9]{3})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])?$",
+                },
+            },
+            "required": ["accession_number"],
+            "additionalProperties": False,
+        })
+    >>> table.add_row(metadata={"accession_number": 1234, "collection_date": "2020-01-01"})
+    >>> table.metadata
+    array([123,  34,  97,  99,  99, 101, 115, 115, 105, 111, 110,  95, 110,
+       117, 109,  98, 101, 114,  34,  58,  32,  49,  50,  51,  52,  44,
+        32,  34,  99, 111, 108, 108, 101,  99, 116, 105, 111, 110,  95,
+       100,  97, 116, 101,  34,  58,  32,  34,  50,  48,  50,  48,  45,
+        48,  49,  45,  48,  49,  34, 125], dtype=int8)
+    >>> table[0].metadata
+    {'accession_number': 1234, 'collection_date': '2020-01-01'}
+
+Note that now ``add_row`` can now be passed an object and ``t[0].metadata`` returns an
+object.
+
+In the interests of efficiency, the table ``set_columns`` and
+``append_columns`` methods still require binary encoded metadata and offsets,
+which can be prepared from a list of metadata objects with the following::
+
+    >>> metadata_column = [
+        {"accession_number": 1234, "collection_date": "2020-01-01"},
+        {"accession_number": 1235, "collection_date": "2020-01-02"},
+        {"accession_number": 1236, "collection_date": "2020-01-03"}]
+    >>> encoded_metadata_column = [
+                table.metadata_schema.validate_and_encode_row(r)
+                for r in metadata_column
+        ]
+    >>> metadata, metadata_offset = tskit.util.pack_bytes(encoded_metadata_column)
```

Comment:
I think you can just call ``tskit.pack_bytes`` - we should discourage users from knowing what the internal package layout is (because we want to be able to change this in the future).

Reply:
Done in https://github.com/tskit-dev/tskit/pull/543/commits/eaae12304944666fc909b8584fd15c09ea4ab6a7